src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,217 @@ func (s *dbShard) Repair(
 	}
-	defer reader.Close()
 
-	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
-	bytesReader := bytes.NewReader(nil)
-	dataPointIter := m3tsz.NewReaderIterator(bytesReader, m3tsz.DefaultIntOptimizationEnabled, encodingOpts)
-	var lastWriteError error
-	var processedBlockCount int64
+	var (
+		processedTileCount  atomic.Int64
+		multiErr            xerrors.MultiError
+		dp                  ts.Datapoint
+		segmentCapacity     int
+		writerData          = make([][]byte, 2)
+	)
+
+	for readerIter.Next() {
+		seriesIter, id, encodedTags := readerIter.Current()
 
-	for {
-		id, tags, data, _, err := reader.Read()
-		if err == io.EOF {
-			break
+		for seriesIter.Next() {
+			frame := seriesIter.Current()
+			unit, singleUnit := frame.Units().SingleValue()
+			annotation, singleAnnotation := frame.Annotations().SingleValue()
+
+			if frameValues := frame.Values(); len(frameValues) > 0 {
+				lastIdx := len(frameValues) - 1
+
+				value := frameValues[lastIdx]
+				timestamp := frame.Timestamps()[lastIdx]
+
+				if !singleUnit {
+					// TODO: what happens if unit has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Units().Values()) {
+						unit = frame.Units().Values()[lastIdx]
+					}
+				}
+
+				if !singleAnnotation {
+					// TODO: what happens if annotation has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Annotations().Values()) {
+						annotation = frame.Annotations().Values()[lastIdx]
+					}
+				}
+
+				dp.Timestamp = timestamp
+				dp.Value = value
+				err = encoder.Encode(dp, unit, annotation)
+				if err != nil {
+					s.metrics.largeTilesWriteErrors.Inc(1)
+					multiErr = multiErr.Add(err)
+				}
+
+				processedTileCount.Inc()
+			}
 		}
-		if err != nil {
-			return processedBlockCount, err
+
+		segment := encoder.DiscardReset(opts.Start, segmentCapacity, targetSchemaDescr)
+
+		segmentLen := segment.Len()
+		if segmentLen > segmentCapacity {
+			// Will use the same capacity for the next series.
+			segmentCapacity = segmentLen
 		}
 
-		data.IncRef()
-		bytesReader.Reset(data.Bytes())
-		dataPointIter.Reset(bytesReader, nil)
+		writerData[0] = segment.Head.Bytes()
+		writerData[1] = segment.Tail.Bytes()
+		checksum := segment.CalculateChecksum()
 
-		for dataPointIter.Next() {
-			dp, unit, annot := dataPointIter.Current()
-			_, err = s.writeAndIndex(ctx, id, tags, dp.Timestamp, dp.Value, unit, annot, wOpts, true)
-			if err != nil {
-				s.metrics.largeTilesWriteErrors.Inc(1)
-				lastWriteError = err
-			}
+		if err := writer.WriteAll(id.Bytes(), encodedTags, writerData, checksum); err != nil {
+			s.metrics.largeTilesWriteErrors.Inc(1)
+			multiErr = multiErr.Add(err)
+		} else {
+			s.metrics.largeTilesWrites.Inc(1)
 		}
 
-		dataPointIter.Close()
+		id.Finalize()
+		//FIXME apparently this does not finalize segment.Tail as FinalizeTail flag is not set by default.
```

Comment:
This is ok though because tails are not pooled and come from a fixed constant allocated arrays inside the encoding package.

Can remove this comment.

Reply:
Happy to hear this because I struggled with finding a good solution for it.